packages/thirdweb/src/react/core/hooks/contract/useSendTransaction.ts

```diff
@@ -97,34 +98,66 @@ export function useSendTransactionCore(
               return;
             }
 
-            //  buy supported, check if there is enough balance - if not show modal to buy tokens
-            const [walletBalance, totalCostWei] = await Promise.all([
-              getWalletBalance({
+            if (!tx.valueERC20) {
+              //  buy supported, check if there is enough balance - if not show modal to buy tokens
+              const [walletBalance, totalCostWei] = await Promise.all([
+                getWalletBalance({
+                  address: account.address,
+                  chain: tx.chain,
+                  client: tx.client,
+                }),
+                getTotalTxCostForBuy(tx, account?.address),
+              ]);
+
+              const walletBalanceWei = walletBalance.value;
+
+              // if enough balance, send tx
+              if (totalCostWei < walletBalanceWei) {
+                sendTx();
+                return;
+              }
+
+              // if not enough balance - show modal
+              showPayModal({
+                tx,
+                sendTx,
+                rejectTx: () => {
+                  reject(new Error("Not enough balance"));
+                },
+                cost: toEther(totalCostWei),
+                walletBalance,
+              });
+            }
+
+            // overridden cost
+            else {
+              const walletBalance = await getWalletBalance({
                 address: account.address,
                 chain: tx.chain,
                 client: tx.client,
-              }),
-              getTotalTxCostForBuy(tx, account?.address),
-            ]);
-
-            const walletBalanceWei = walletBalance.value;
-
-            // if enough balance, send tx
-            if (totalCostWei < walletBalanceWei) {
-              sendTx();
-              return;
+                tokenAddress: tx.valueERC20.tokenAddress,
+              });
+
+              // if enough balance, send tx
```

Comment:
we might need to check approval too maybe?

what's cool with this API is that we can automatically prompt for approval if needed, for any transaction

Reply:
It feels weird to send an approval transaction automatically before the given transaction. UX would be confusing for mobile users because they would be prompted to do 2 transactions one after another which would suddenly send them to wallet app. I think we should leave that up to developer so they can design the 2-step UI ( similar to what we have in Pay UI )